src/plugins/minimal_scene/MinimalScene.cc

```diff
@@ -857,30 +935,57 @@ QSGNode *RenderWindowItem::updatePaintNode(QSGNode *_node,
 {
   TextureNode *node = static_cast<TextureNode *>(_node);
 
-  if (!this->dataPtr->renderThread->context)
+  if (!this->dataPtr->initialized)
   {
-    QOpenGLContext *current = this->window()->openglContext();
-    // Some GL implementations require that the currently bound context is
-    // made non-current before we set up sharing, so we doneCurrent here
-    // and makeCurrent down below while setting up our own context.
-    current->doneCurrent();
-
-    this->dataPtr->renderThread->context = new QOpenGLContext();
-    this->dataPtr->renderThread->context->setFormat(current->format());
-    this->dataPtr->renderThread->context->setShareContext(current);
-    this->dataPtr->renderThread->context->create();
-    this->dataPtr->renderThread->context->moveToThread(
-        this->dataPtr->renderThread);
+    // Exit immediately if still initializing
+    if (this->dataPtr->initializing)
+    {
+      return nullptr;
+    }
+    this->dataPtr->initializing = true;
+
+    // Set the render thread's render system
+    this->dataPtr->renderThread->SetGraphicsAPI(
+        this->dataPtr->graphicsAPI);
+
+    if (this->dataPtr->graphicsAPI == rendering::GraphicsAPI::OPENGL)
+    {
+      QOpenGLContext *current = this->window()->openglContext();
+      // Some GL implementations require that the currently bound context is
+      // made non-current before we set up sharing, so we doneCurrent here
+      // and makeCurrent down below while setting up our own context.
+      current->doneCurrent();
+
+      this->dataPtr->renderThread->SetContext(new QOpenGLContext());
+      this->dataPtr->renderThread->Context()->setFormat(current->format());
+      this->dataPtr->renderThread->Context()->setShareContext(current);
+      this->dataPtr->renderThread->Context()->create();
+
+      // The slot "Ready" runs on the main thread, move the context to match
+      this->dataPtr->renderThread->Context()->moveToThread(
+          QApplication::instance()->thread());
 
-    current->makeCurrent(this->window());
+      current->makeCurrent(this->window());
 
-    QMetaObject::invokeMethod(this, "Ready");
+      // Initialize on main thread
+      QMetaObject::invokeMethod(this, "Ready", Qt::QueuedConnection);
+    }
+    else if (this->dataPtr->graphicsAPI == rendering::GraphicsAPI::METAL)
+    {
+      // Initialize on main thread
+      QMetaObject::invokeMethod(this, "Ready", Qt::QueuedConnection);
+    }
+    else
+    {
+      // invalid render system
```

Comment:
print an error msg?

Reply:
Updated in ca52977561c63bb615e01b25fa62b350645dfda6